File: plugins/plugin-linear/typescript/src/actions/getActivity.ts
import {
  type Action,
  type ActionResult,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  ModelType,
  type State,
} from "@elizaos/core";
import { getActivityTemplate } from "../generated/prompts/typescript/prompts.js";
import type { LinearService } from "../services/linear";

export const getActivityAction: Action = {
  name: "GET_LINEAR_ACTIVITY",
  description: "Get recent Linear activity log with optional filters",
  similes: [
    "get-linear-activity",
    "show-linear-activity",
    "view-linear-activity",
    "check-linear-activity",
  ],

  examples: [
    [
      {
        name: "User",
        content: {
          text: "Show me recent Linear activity",
        },
      },
      {
        name: "Assistant",
        content: {
          text: "I'll show you the recent Linear activity.",
          actions: ["GET_LINEAR_ACTIVITY"],
        },
      },
    ],
    [
      {
        name: "User",
        content: {
          text: "What happened in Linear today?",
        },
      },
      {
        name: "Assistant",
        content: {
          text: "Let me check today's Linear activity for you.",
          actions: ["GET_LINEAR_ACTIVITY"],
        },
      },
    ],
    [
      {
        name: "User",
        content: {
          text: "Show me what issues John created this week",
        },
      },
      {
        name: "Assistant",
        content: {
          text: "I'll find the issues John created this week.",
          actions: ["GET_LINEAR_ACTIVITY"],
        },
      },
    ],
  ],

  async validate(runtime: IAgentRuntime, _message: Memory, _state?: State): Promise<boolean> {
    const apiKey = runtime.getSetting("LINEAR_API_KEY");
    return !!apiKey;
  },

  async handler(
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult> {
    try {
      const linearService = runtime.getService<LinearService>("linear");
      if (!linearService) {
        throw new Error("Linear service not available");
      }

      const content = message.content.text || "";
      const filters: Record<string, unknown> = {};
      let limit = 10;

      if (content) {
        const prompt = getActivityTemplate.replace("{{userMessage}}", content);
        const response = await runtime.useModel(ModelType.TEXT_LARGE, {
          prompt: prompt,
        });

        if (response) {
          try {
            const parsed = JSON.parse(
              response
                .replace(/^```(?:json)?\n?/, "")
                .replace(/\n?```$/, "")
                .trim()
            );

            if (parsed.timeRange) {
              const now = new Date();
              let fromDate: Date | undefined;

              if (parsed.timeRange.from) {
                fromDate = new Date(parsed.timeRange.from);
              } else if (parsed.timeRange.period) {
                switch (parsed.timeRange.period) {
                  case "today":
                    fromDate = new Date(now.setHours(0, 0, 0, 0));
                    break;
                  case "yesterday":
                    fromDate = new Date(now.setDate(now.getDate() - 1));
                    fromDate.setHours(0, 0, 0, 0);
                    break;
                  case "this-week":
                    fromDate = new Date(now.setDate(now.getDate() - now.getDay()));
                    fromDate.setHours(0, 0, 0, 0);
                    break;
                  case "last-week":
                    fromDate = new Date(now.setDate(now.getDate() - now.getDay() - 7));
                    fromDate.setHours(0, 0, 0, 0);
                    break;
                  case "this-month":
                    fromDate = new Date(now.getFullYear(), now.getMonth(), 1);
                    break;
                }
              }

              if (fromDate) {
                filters.fromDate = fromDate.toISOString();
              }
            }

            if (parsed.actionTypes && parsed.actionTypes.length > 0) {
              filters.action = parsed.actionTypes[0];
            }

            if (parsed.resourceTypes && parsed.resourceTypes.length > 0) {
              filters.resource_type = parsed.resourceTypes[0];
            }

            if (parsed.resourceId) {
              filters.resource_id = parsed.resourceId;
            }

            if (parsed.successFilter && parsed.successFilter !== "all") {
              filters.success = parsed.successFilter === "success";
            }

            limit = parsed.limit || 10;
          } catch (parseError) {
            logger.warn("Failed to parse activity filters:", parseError);
          }
        }
      }

      let activity = linearService.getActivityLog(limit * 2, filters);

      if (filters.fromDate) {
        const fromDateValue = filters.fromDate;
        const fromDate =
          typeof fromDateValue === "string"
            ? fromDateValue
            : fromDateValue instanceof Date
              ? fromDateValue.toISOString()
              : String(fromDateValue);
        const fromTime = new Date(fromDate).getTime();
        if (!Number.isNaN(fromTime)) {
          activity = activity.filter((item) => new Date(item.timestamp).getTime() >= fromTime);
        }
      }

      activity = activity.slice(0, limit);

      if (activity.length === 0) {
        const noActivityMessage = filters.fromDate
          ? `No Linear activity found for the specified filters.`
          : "No recent Linear activity found.";
        await callback?.({
          text: noActivityMessage,
          source: message.content.source,
        });
        return {
          text: noActivityMessage,
          success: true,
          data: {
            activity: [],
          },
        };
      }

      const activityText = activity
        .map((item, index) => {
          const time = new Date(item.timestamp).toLocaleString();
          const status = item.success ? "✅" : "❌";
          const details = Object.entries(item.details)
            .filter(([key]) => key !== "filters")
            .map(([key, value]) => `${key}: ${JSON.stringify(value)}`)
            .join(", ");

          return `${index + 1}. ${status} ${item.action} on ${item.resource_type} ${item.resource_id}\n   Time: ${time}\n   ${details ? `Details: ${details}` : ""}${item.error ? `\n   Error: ${item.error}` : ""}`;
        })
        .join("\n\n");

      const headerText = filters.fromDate
        ? `📊 Linear activity ${content}:`
        : "📊 Recent Linear activity:";

      const resultMessage = `${headerText}\n\n${activityText}`;
      await callback?.({
        text: resultMessage,
        source: message.content.source,
      });

      return {
        text: `Found ${activity.length} activity item${activity.length === 1 ? "" : "s"}`,
        success: true,
        data: {
          activity: activity.map((item) => ({
            id: item.id,
            action: item.action,
            resource_type: item.resource_type,
            resource_id: item.resource_id,
            success: item.success,
            error: item.error,
            details: JSON.stringify(item.details) as string,
            timestamp:
              typeof item.timestamp === "string"
                ? item.timestamp
                : new Date(item.timestamp).toISOString(),
          })) as Array<Record<string, string | boolean | undefined>>,
          filters: filters
            ? {
                ...filters,
                fromDate: filters.fromDate
                  ? typeof filters.fromDate === "string"
                    ? filters.fromDate
                    : String(filters.fromDate)
                  : undefined,
              }
            : undefined,
          count: activity.length,
        },
      };
    } catch (error) {
      logger.error("Failed to get activity:", error);
      const errorMessage = `❌ Failed to get activity: ${error instanceof Error ? error.message : "Unknown error"}`;
      await callback?.({
        text: errorMessage,
        source: message.content.source,
      });
      return {
        text: errorMessage,
        success: false,
      };
    }
  },
};

File: plugins/plugin-linear/python/src/elizaos_plugin_linear/actions/get_activity.py
import json
import logging
import re
from datetime import datetime, timedelta
from typing import Any

from elizaos_plugin_linear.actions.base import (
    ActionExample,
    ActionResult,
    HandlerCallback,
    Memory,
    RuntimeProtocol,
    State,
    create_action,
)
from elizaos_plugin_linear.services.linear import LinearService

logger = logging.getLogger(__name__)

GET_ACTIVITY_TEMPLATE = """Extract activity filter criteria from the user's request.

User request: "{user_message}"

Return ONLY a JSON object:
{{
  "timeRange": {{
    "period": "today/yesterday/this-week/last-week/this-month"
  }},
  "actionTypes": ["create_issue/update_issue/delete_issue/create_comment/etc"],
  "resourceTypes": ["issue/project/comment/team"],
  "resourceId": "Specific resource ID if mentioned (e.g., ENG-123)",
  "successFilter": "success/failed/all",
  "limit": number (default 10)
}}

Only include fields that are clearly mentioned."""


async def validate(
    runtime: RuntimeProtocol,
    _message: Memory,
    _state: State | None = None,
) -> bool:
    try:
        api_key = runtime.get_setting("LINEAR_API_KEY")
        return bool(api_key)
    except Exception:
        return False


async def handler(
    runtime: RuntimeProtocol,
    message: Memory,
    _state: State | None = None,
    options: dict[str, Any] | None = None,
    callback: HandlerCallback | None = None,
) -> ActionResult:
    try:
        linear_service: LinearService = runtime.get_service("linear")
        if not linear_service:
            raise RuntimeError("Linear service not available")

        content = message.get("content", {}).get("text", "")
        filters: dict[str, Any] = {}
        limit = 10

        if content:
            prompt = GET_ACTIVITY_TEMPLATE.format(user_message=content)
            response = await runtime.use_model("TEXT_LARGE", {"prompt": prompt})

            if response:
                try:
                    cleaned = re.sub(r"^```(?:json)?\n?", "", response)
                    cleaned = re.sub(r"\n?```$", "", cleaned).strip()
                    parsed = json.loads(cleaned)

                    if parsed.get("timeRange"):
                        period = parsed["timeRange"].get("period")
                        now = datetime.now()
                        from_date: datetime | None = None

                        if period == "today":
                            from_date = now.replace(hour=0, minute=0, second=0, microsecond=0)
                        elif period == "yesterday":
                            from_date = (now - timedelta(days=1)).replace(
                                hour=0, minute=0, second=0, microsecond=0
                            )
                        elif period == "this-week":
                            from_date = (now - timedelta(days=now.weekday())).replace(
                                hour=0, minute=0, second=0, microsecond=0
                            )
                        elif period == "last-week":
                            from_date = (now - timedelta(days=now.weekday() + 7)).replace(
                                hour=0, minute=0, second=0, microsecond=0
                            )
                        elif period == "this-month":
                            from_date = now.replace(
                                day=1, hour=0, minute=0, second=0, microsecond=0
                            )

                        if from_date:
                            filters["fromDate"] = from_date.isoformat()

                    if parsed.get("actionTypes"):
                        filters["action"] = parsed["actionTypes"][0]

                    if parsed.get("resourceTypes"):
                        filters["resource_type"] = parsed["resourceTypes"][0]

                    if parsed.get("resourceId"):
                        filters["resource_id"] = parsed["resourceId"]

                    if parsed.get("successFilter") and parsed["successFilter"] != "all":
                        filters["success"] = parsed["successFilter"] == "success"

                    limit = parsed.get("limit", 10)

                except json.JSONDecodeError:
                    logger.warning("Failed to parse activity filters")

        activity = linear_service.get_activity_log(limit * 2, filters if filters else None)

        if filters.get("fromDate"):
            from_time = datetime.fromisoformat(filters["fromDate"])
            activity = [
                item
                for item in activity
                if datetime.fromisoformat(
                    item.timestamp.replace("Z", "+00:00").replace("+00:00", "")
                )
                >= from_time
            ]

        activity = activity[:limit]

        if not activity:
            no_activity_msg = (
                "No Linear activity found for the specified filters."
                if filters.get("fromDate")
                else "No recent Linear activity found."
            )
            if callback:
                await callback(
                    {"text": no_activity_msg, "source": message.get("content", {}).get("source")}
                )
            return {"text": no_activity_msg, "success": True, "data": {"activity": []}}

        activity_text = []
        for i, item in enumerate(activity):
            time_str = datetime.fromisoformat(item.timestamp.replace("Z", "")).strftime(
                "%Y-%m-%d %H:%M"
            )
            status = "✅" if item.success else "❌"

            details_str = ", ".join(f"{k}: {v}" for k, v in item.details.items() if k != "filters")

            error_line = f"\n   Error: {item.error}" if item.error else ""
            details_line = f"Details: {details_str}" if details_str else ""
            activity_text.append(
                f"{i + 1}. {status} {item.action} on {item.resource_type} {item.resource_id}\n"
                f"   Time: {time_str}\n"
                f"   {details_line}{error_line}"
            )

        header_text = (
            f"📊 Linear activity {content}:"
            if filters.get("fromDate")
            else "📊 Recent Linear activity:"
        )

        result_message = f"{header_text}\n\n" + "\n\n".join(activity_text)
        if callback:
            await callback(
                {"text": result_message, "source": message.get("content", {}).get("source")}
            )

        return {
            "text": f"Found {len(activity)} activity item{'s' if len(activity) != 1 else ''}",
            "success": True,
            "data": {
                "activity": [
                    {
                        "id": item.id,
                        "timestamp": item.timestamp,
                        "action": item.action,
                        "resource_type": item.resource_type,
                        "resource_id": item.resource_id,
                        "success": item.success,
                        "error": item.error,
                    }
                    for item in activity
                ],
                "filters": filters,
                "count": len(activity),
            },
        }

    except Exception as error:
        logger.error(f"Failed to get activity: {error}")
        error_message = f"❌ Failed to get activity: {error}"
        if callback:
            await callback(
                {"text": error_message, "source": message.get("content", {}).get("source")}
            )
        return {"text": error_message, "success": False}


get_activity_action = create_action(
    name="GET_LINEAR_ACTIVITY",
    description="Get recent Linear activity log with optional filters",
    similes=["get-linear-activity", "show-linear-activity", "view-linear-activity"],
    examples=[
        [
            ActionExample(name="User", content={"text": "Show me recent Linear activity"}),
            ActionExample(
                name="Assistant",
                content={
                    "text": "I'll show you the recent Linear activity.",
                    "actions": ["GET_LINEAR_ACTIVITY"],
                },
            ),
        ],
    ],
    validate=validate,
    handler=handler,
)
